returnn/torch/engine.py

```diff
@@ -177,11 +210,35 @@ def train_epoch(self):
         accumulated_losses_dict = NumbersDict()
         accumulated_inv_norm_factors_dict = NumbersDict()
         step_idx = 0
-        for data in self._train_dataloader:
-            self._updater.get_optimizer().zero_grad()
+        epoch_start_time = time.time()
+
+        data_iter = iter(self._train_dataloader)
+        elapsed_computation_time = 0
+
+        while True:
+            data = next(data_iter, None)
+            _has_data = torch.tensor([data is not None], dtype=torch.int8).to(self._device)
```

Comment:
if I do not use `to(self._device)`, I get the error RuntimeError: Tensors must be CUDA and dense

Reply:
Where do you get that error?

Reply:
from here `torch.distributed.all_reduce(_has_data, op=torch.distributed.ReduceOp.MIN)`

Reply:
Oh ok. I did not expect that it can only use one single mechanism for communication, namely NCCL here. This might be a bit suboptimal here (roundtrip CPU -> GPU -> NCCL -> GPU -> CPU; I was rather expecting CPU -> ethernet -> CPU). But some big warning comment then, and maybe we later need to deal with this.

Reply:
```suggestion
            # WARNING: torch.distributed works only for the registered device,
            # as it uses only one mechanism for communication, like NCCL.
            # This is suboptimal here as we have the roundtrip CPU -> GPU -> NCCL -> GPU -> CPU.
            # TODO: Use more direct CPU -> Ethernet -> CPU communication.
            _has_data = torch.tensor([data is not None], dtype=torch.int8).to(self._device)
```

Reply:
okay, I have added the warning comment here